doc_src/cmds/for.rst

```diff
@@ -15,6 +15,8 @@ Description
 
 ``for`` is a loop construct. It will perform the commands specified by ``COMMANDS`` multiple times. On each iteration, the local variable specified by ``VARNAME`` is assigned a new value from ``VALUES``. If ``VALUES`` is empty, ``COMMANDS`` will not be executed at all. The ``VARNAME`` is visible when the loop terminates and will contain the last value assigned to it. If ``VARNAME`` does not already exist it will be set in the local scope. For our purposes if the ``for`` block is inside a function there must be a local variable with the same name. If the ``for`` block is not nested inside a function then global and universal variables of the same name will be used if they exist.
 
+The exit status of the for loop is the exit status of the last iteration of the ``COMMANDS`` executed, or 0 if none were executed.
```

Comment:
The "or 0 if none were executed" is not always true, for example

	fish -c 'false; for a in; end' # exits with 1.

So the `for`-statement itself does not set an exit status, unlike `while`:

	fish -c 'sh -c "exit 2"; while false; end' # exits with 0

We could drop this commit, because it should go without saying that the
last-executed command sets the exit status.

Reply:
> So the for-statement itself does not set an exit status, unlike while:

Is it the inconsistence must be resolved or it is intended behavior and why? (Should I create new issue about it?)

Reply:
It would be part two of this: https://github.com/fish-shell/fish-shell/issues/4982

Reply:
Ah right. I believe we should resolve the inconsistency to match other shells. The docs for `while` say

> The exit status of the while loop is the exit status of the last iteration of the ``COMMANDS`` executed, or 0 if none were executed. (This matches other shells and is POSIX-compatible.)

Turns out `for` loops are treated the same way in other shells, try:

    sh -c 'false; for x in; do :; done; echo $?' # prints 0

So we should probably just teach `run_for_statement()` to set the exit code to zero if no loop iteration was executed.
This matches both [bash](https://www.gnu.org/software/bash/manual/html_node/Looping-Constructs.html#Looping-Constructs) and [POSIX](https://pubs.opengroup.org/onlinepubs/9699919799/utilities/V3_chap02.html).

No need to open an issue - we don't get paid in issues ;) - having the discussion here should be enough.

Reply:
I like it @krobelus.

Reply:
The fix to set $status to 0 when no loop iteration is executed seems okay but there's another edge case with empty for-loop bodies.
This loop doesn't change $status

	false
	for x in 1 2 3
	end
	# status is 1

which is weird when comparing to `for x in; end` which does (with the fix)

I guess we can just ignore this case because an empty body in a for-loop makes no sense. We could even error in that case..

Reply:
ok nevermind, `for` is just like `set` -- it doesn't touch the exit status by itself.
Documented in 0acbbfe2ca7847f8d01e36b17f7e78336d08e5c7